### DISCUSSION TITLE
Static in Production for fastapi.

### DISCUSSION THREAD

**Discussion by philipokiokio** [aminalaee/sqladmin#221] [Q&A] [answered]
Hi, So i am temporarily using sqladmin in my fast API application in production, but the static portion of the application is not working in production unlike in development (local computer). 

How can I fix this?

**Comment by aminalaee** [comment]
Hey,
Do you mean the static files kike CSS and JS are not loading? How do you run your app?

**Reply by philipokiokio** [reply]
Hi Aminalee.


I mean that static files are not available in production, but they are available in development. 

uvicorn main:app

**Reply by philipokiokio** [reply]
So the Static files are not rendering in production.

**Reply by aminalaee** [reply]
TBH unless you provide more information about your deployment I don't know how to help, you can see for example here I've deployed this project: https://github.com/aminalaee/sqladmin-demo at https://python-sqladmin.herokuapp.com/admin/

I'm guessing it might be a web server like Nginx config.

**Reply by ilnuru** [reply]
Does nginx need additional configuration?

**Reply by aminalaee** [reply]
It's possible to override some paths with them, do you also face the same issue?

**Reply by ilnuru** [reply]
Yes, I have the same issue.

**Reply by ilnuru** [reply]
We use kubernetes for deployment. There is an Ingress controller with nginx

**Reply by ilnuru** [reply]
Found the problem.
The templates generate http links, but the prod server uses https. Static files availible with https scheme.
In my case, it was enough to start uvicorn with the --proxy-headers flag. Now uvicorn is getting the https usage headers, but url_for in templates not using this. <link rel="stylesheet" href="{{ url_for('admin:statics', path='css/tabler.min.css') }}"> generate http://devserver.url

More info on this issue https://github.com/encode/starlette/issues/538

**Reply by ilnuru** [reply]
Flag --forwarded-allow-ips * resolve this problem. Now starlette detects https correctly.
In the flag, you can specify * or the necessary ip. https://github.com/encode/uvicorn/blob/master/uvicorn/middleware/proxy_headers.py

**Reply by aminalaee** [reply]
Great, thanks for taking a look into this @ilnuru 
This has come up before, but I thought this is different.

Can you also post these as a separate comment so I can mark as answer?

**Comment by ilnuru** [comment, marked as answer]
The probable reason is that the static is connected via http, and your site is opened via http**s**.
To solve this, you need to make sure that your proxy server (like nginx ) is passing the necessary headers to your application. https://github.com/encode/starlette/issues/538#issuecomment-780608541
Next, you need to configure uvicorn. The `--forwarded-allow-ips` flag tells the server which ips to accept these headers from. The `--proxy-headers` flag will enable passing the required headers. [https://www.uvicorn.org/deployment/](url)
Now inside your fastapi or starlette app `scope["scheme"]` contains `https` and the app will generate a static url based on scheme.

**Comment by stefanosmav** [comment]
What if there is no proxy server in between the client and the application? My login form and all of the statics in general do not work in production because of the issue that @philipokiokio described. I am using uvicorn via a dockerfile:
_CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "80"]_

**Reply by aminalaee** [reply]
@stefanosmav Does it work on your local env?

**Reply by stefanosmav** [reply]
@aminalaee yes. Everything works locally. But in production (https) I cannot even login in the admin page and the statics don't work.

**Reply by aminalaee** [reply]
@stefanosmav  And how do you run https? With uvicorn?

**Reply by stefanosmav** [reply]
@aminalaee I run https via azure web app. It redirects the port automatically

**Reply by aminalaee** [reply]
@stefanosmav That's very specific and I don't have any experience with that, but if it works on local then I guess something is wrong with your setup and all I can do is forward you to: https://www.uvicorn.org/deployment/#running-behind-nginx
